src/H5Tconv.c

```diff
@@ -7976,8 +8298,11 @@ H5T__conv_f_i(hid_t src_id, hid_t dst_id, H5T_cdata_t *cdata, size_t nelmts, siz
             /* Allocate enough space for the buffer holding temporary
              * converted value
              */
-            buf_size = (size_t)(pow(2.0, (double)src.u.f.esize) / 8 + 1);
-            int_buf  = (uint8_t *)H5MM_calloc(buf_size);
+            if (dst.prec / 8 > src_p->shared->size)
+                buf_size = (dst.prec + 7) / 8;
+            else
+                buf_size = src_p->shared->size;
```

Comment:
This function was previously assuming ~32-bit floats. 16-bit and smaller floats would cause an invalid memory access further on when assuming it could copy more than 16 bits out of the temporary source buffer into the destination.